File: Haskell/ES Feb2016/DataStructures/Graph/EulerianCycle.hs
-------------------------------------------------------------------------------
-- Student's name:
-- Student's group:
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module DataStructures.Graph.EulerianCycle(isEulerian, eulerianCycle) where

import DataStructures.Graph.Graph
import Data.List

--H.1)
isEulerian :: Eq a => Graph a -> Bool
isEulerian g = not (isEmpty g)
isEulerian g = foldr (\e x -> x&&((mod (degree g e) 2) == 0)) True (vertices g)

-- H.2)
remove :: (Eq a) => Graph a -> (a,a) -> Graph a
remove g (v,u) = removeAux(deleteEdge g (v,u))


removeAux g = foldr (\e x -> if ((degree g e)==0) then (deleteVertex x e) else x) g (vertices g)

-- H.3)
extractCycle :: (Eq a) => Graph a -> a -> (Graph a, Path a)
extractCycle g v0 = extractAux g v0 [v0]

extractAux g v l
    | u == head l = (g', p')
    | otherwise = extractAux g' u p'
        where 
            g' = remove g (v,u)
            u = head (successors g v)
            p' = l++[u]
 

-- H.4)
connectCycles :: (Eq a) => Path a -> Path a -> Path a
connectCycles [] ys = ys
connectCycles (x:xs) (y:ys)
    | x == y = (x:ys)++xs
    | otherwise = x:(connectCycles xs (y:ys))

-- H.5)
vertexInCommon :: Eq a => Graph a -> Path a -> a
vertexInCommon g cycle = vertexAux (vertices g) cycle

vertexAux (x:xs) ys
    | elem x ys = x
    | otherwise = vertexAux xs ys

-- H.6) 
eulerianCycle :: Eq a => Graph a -> Path a
eulerianCycle g
    | isEulerian g = eulerianAux g' p'
    | otherwise = error "Grafo meh"
            where
                (g', p') = extractCycle g (head (vertices g))  

eulerianAux g cycle
    | isEmpty g = cycle
    | otherwise = eulerianAux g' (connectCycles cycle p')
        where 
            (g', p') = extractCycle g (vertexInCommon g cycle)

File: Java/ES Feb2016/src/dataStructures/graph/EulerianCycle.java
/**
 * Student's name:
 * Student's group:
 * <p>
 * Data Structures. Grado en Informática. UMA.
 */

package dataStructures.graph;

import dataStructures.list.*;

public class EulerianCycle<V> {
    private List<V> eCycle;

    @SuppressWarnings("unchecked")
    public EulerianCycle(Graph<V> g) {
        Graph<V> graph = (Graph<V>) g.clone();
        eCycle = eulerianCycle(graph);
    }

    public boolean isEulerian() {
        return eCycle != null;
    }

    public List<V> eulerianCycle() {
        return eCycle;
    }

    // J.1
    private static <V> boolean isEulerian(Graph<V> g) {
        boolean verticesPares = true;

        for (V vertex : g.vertices()) {
            if (g.degree(vertex) % 2 != 0) {
                verticesPares = false;
            }
        }
        return verticesPares;
    }

    // J.2
    private static <V> void remove(Graph<V> g, V v, V u) {
        g.deleteEdge(v, u);

        for (V vertex : g.vertices()) {
            if (g.degree(vertex) == 0) {
                g.deleteVertex(vertex);
            }
        }
    }

    // J.3
    private static <V> List<V> extractCycle(Graph<V> g, V v0) {
        List<V> ciclo = new LinkedList<>();
        ciclo.append(v0);
        V v = v0;

        while (!g.successors(v).isElem(v0) && g.successors(v).iterator().hasNext()){
            V u = g.successors(v).iterator().next();
            ciclo.append(u);
            remove(g, v, u);
            v = u;
        }

        remove(g, v, v0);
        ciclo.append(v0);

        return ciclo;
    }

    // J.4
    private static <V> void connectCycles(List<V> xs, List<V> ys) {
        if (xs.isEmpty()) {
            xs = ys;
        } else {
            int i = 0;
            int j = 1;

            while (!xs.get(i).equals(ys.get(0))) {
                i++;
            }
            i++;

            while (j < ys.size()) {
                xs.insert(i, ys.get(j));
                i++;
                j++;
            }
        }
    }

    // J.5
    private static <V> V vertexInCommon(Graph<V> g, List<V> cycle) {
        V v = cycle.get(0);
        int i = 0;
        boolean enc = false;

        while (i < cycle.size() && !enc) {
            if (g.vertices().isElem(cycle.get(i))) {
                v = cycle.get(i);
                enc = true;
            }
            i++;
        }


        return v;
    }

    // J.6
    private static <V> List<V> eulerianCycle(Graph<V> g) {
        if (isEulerian(g)) {
            V v;
            List<V> ciclo = extractCycle(g, g.vertices().iterator().next());

            while (!g.isEmpty()) {
                v = vertexInCommon(g, ciclo);
                var cicloAux = extractCycle(g, v);
                connectCycles(ciclo, cicloAux);
            }

            return ciclo;
        }


        return null;
    }
}
